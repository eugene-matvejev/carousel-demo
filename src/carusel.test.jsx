import React from 'react';
import Carusel from './carusel';
import { shallow } from 'enzyme';

describe('<Carusel/>', () => {
    const props = {
        items: [
            {
                src: '{{src0}}',
                tags: '{{tags0}}',
            },
            {
                src: '{{src1}}',
                tags: '{{tags1}}',
            },
        ],
    };

    describe('render', () => {
        it('with required/default props', () => {
            const c = shallow(<Carusel {...props} />);

            expect(c).toMatchSnapshot();
        });

        describe('with optional props', () => {
            [
                ['qaSelector', '{{qaSelector}}'],
                ['title', '{{title}}'],
                ['position', 1],
                ['limit', 1],
            ].forEach(([prop, v]) => {
                it(`::${prop} as "${v}"`, () => {
                    const c = shallow(<Carusel {...props} {...{ [prop]: v }} />);

                    expect(c).toMatchSnapshot();
                });
            });
        });
    });

    describe('internal callbacks', () => {
        describe('::handleNewPosition', () => {
            it('when [::limit] smaller than a distance from [::position] till end of the queue it should present elements in queue between [::position] and [::position + ::limit]', () => {
                const items = [
                    {
                        src: '{{src0}}',
                        tags: '{{tags0}}',
                    },
                    {
                        src: '{{src1}}',
                        tags: '{{tags1}}',
                    },
                    {
                        src: '{{src2}}',
                        tags: '{{tags2}}',
                    },
                    {
                        src: '{{src3}}',
                        tags: '{{tags3}}',
                    },
                ];
                const c = shallow(<Carusel {...props} items={items} limit={2} position={1} />);

                expect(c.state('items')).toMatchSnapshot();
            });

            it(`when [::limit] higher than a distance from [::position] till end of the queue, it should fill queue by injecting items from start of scope`, () => {
                const c = shallow(<Carusel {...props} limit={3} position={2} />);

                expect(c.state('items')).toMatchSnapshot();
            });
        });

        describe('::onClickLeft', () => {
            it('from click on [data-cy="carusel-controller--left"], should decrease internal state field [::position] by 1, as current position above 0', () => {
                const c = shallow(<Carusel {...props} limit={2} position={1}/>);

                c.find('[data-cy="carusel-controller--left"]').simulate('click');

                expect(c.state('position')).toBe(0);
            });

            it('from click on [data-cy="carusel-controller--left"], should set internal state field [::position] to [items amount in context - 1], as current position is 0', () => {
                const c = shallow(<Carusel {...props} limit={2} />);

                c.find('[data-cy="carusel-controller--left"]').simulate('click');

                expect(c.state('position')).toBe(1);
            });
        });

        describe('::onClickRight', () => {
            it('from click on [data-cy="carusel-controller--right"], should increase internal state field [::position] by 1, as current position is less than an amount of items in context', () => {
                const c = shallow(<Carusel {...props} limit={2} />);

                c.find('[data-cy="carusel-controller--right"]').simulate('click');

                expect(c.state('position')).toBe(1);
            });

            it('from click on [data-cy="carusel-controller--right"], should set internal state field [::position] to 0, as current position is equal to amount of items in context', () => {
                const c = shallow(<Carusel {...props} limit={2} position={1} />);

                c.find('[data-cy="carusel-controller--right"]').simulate('click');

                expect(c.state('position')).toBe(0);
            });

            /** ::componentDidMount fix an issue, obsolete test */
            it.skip('from click on [data-cy="carusel-controller--right"], should set internal state field [::position] to 0, as current position is bigger to amount of items in context', () => {
                const c = shallow(<Carusel {...props} limit={2} position={2} />);

                c.find('[data-cy="carusel-controller--right"]').simulate('click');

                expect(c.state('position')).toBe(0);
            });
        });
    });
});