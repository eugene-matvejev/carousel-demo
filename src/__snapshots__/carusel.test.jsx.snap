// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carusel/> render with optional props ::limit as "1" 1`] = `
<section
  className="carusel"
>
  <div
    className="carusel_container"
  >
    <CaruselItem
      key="0"
      position={0}
      qaSelector="carusel-item__0"
      src="{{src0}}"
      tags="{{tags0}}"
    />
  </div>
  <div
    className="carusel_controls"
  >
    <div
      className="carusel_control--left"
      data-cy="carusel-controller--left"
      onClick={[Function]}
    >
      Prev
    </div>
    <div
      className="carusel_control--right"
      data-cy="carusel-controller--right"
      onClick={[Function]}
    >
      Next
    </div>
  </div>
</section>
`;

exports[`<Carusel/> render with optional props ::position as "1" 1`] = `
<section
  className="carusel"
>
  <div
    className="carusel_container"
  >
    <CaruselItem
      key="0"
      position={1}
      qaSelector="carusel-item__0"
      src="{{src1}}"
      tags="{{tags1}}"
    />
    <CaruselItem
      key="1"
      position={0}
      qaSelector="carusel-item__1"
      src="{{src0}}"
      tags="{{tags0}}"
    />
    <CaruselItem
      key="2"
      position={1}
      qaSelector="carusel-item__2"
      src="{{src1}}"
      tags="{{tags1}}"
    />
  </div>
  <div
    className="carusel_controls"
  >
    <div
      className="carusel_control--left"
      data-cy="carusel-controller--left"
      onClick={[Function]}
    >
      Prev
    </div>
    <div
      className="carusel_control--right"
      data-cy="carusel-controller--right"
      onClick={[Function]}
    >
      Next
    </div>
  </div>
</section>
`;

exports[`<Carusel/> render with optional props ::qaSelector as "{{qaSelector}}" 1`] = `
<section
  className="carusel"
>
  <div
    className="carusel_container"
  >
    <CaruselItem
      key="0"
      position={0}
      qaSelector="{{qaSelector}}carusel-item__0"
      src="{{src0}}"
      tags="{{tags0}}"
    />
    <CaruselItem
      key="1"
      position={1}
      qaSelector="{{qaSelector}}carusel-item__1"
      src="{{src1}}"
      tags="{{tags1}}"
    />
    <CaruselItem
      key="2"
      position={0}
      qaSelector="{{qaSelector}}carusel-item__2"
      src="{{src0}}"
      tags="{{tags0}}"
    />
    <CaruselItem
      key="3"
      position={1}
      qaSelector="{{qaSelector}}carusel-item__3"
      src="{{src1}}"
      tags="{{tags1}}"
    />
  </div>
  <div
    className="carusel_controls"
  >
    <div
      className="carusel_control--left"
      data-cy="{{qaSelector}}carusel-controller--left"
      onClick={[Function]}
    >
      Prev
    </div>
    <div
      className="carusel_control--right"
      data-cy="{{qaSelector}}carusel-controller--right"
      onClick={[Function]}
    >
      Next
    </div>
  </div>
</section>
`;

exports[`<Carusel/> render with optional props ::title as "{{title}}" 1`] = `
<section
  className="carusel"
>
  <div
    className="carusel_title"
    data-cy="carusel-title"
  >
    {{title}}
  </div>
  <div
    className="carusel_container"
  >
    <CaruselItem
      key="0"
      position={0}
      qaSelector="carusel-item__0"
      src="{{src0}}"
      tags="{{tags0}}"
    />
    <CaruselItem
      key="1"
      position={1}
      qaSelector="carusel-item__1"
      src="{{src1}}"
      tags="{{tags1}}"
    />
    <CaruselItem
      key="2"
      position={0}
      qaSelector="carusel-item__2"
      src="{{src0}}"
      tags="{{tags0}}"
    />
    <CaruselItem
      key="3"
      position={1}
      qaSelector="carusel-item__3"
      src="{{src1}}"
      tags="{{tags1}}"
    />
  </div>
  <div
    className="carusel_controls"
  >
    <div
      className="carusel_control--left"
      data-cy="carusel-controller--left"
      onClick={[Function]}
    >
      Prev
    </div>
    <div
      className="carusel_control--right"
      data-cy="carusel-controller--right"
      onClick={[Function]}
    >
      Next
    </div>
  </div>
</section>
`;

exports[`<Carusel/> render with required/default props 1`] = `
<section
  className="carusel"
>
  <div
    className="carusel_container"
  >
    <CaruselItem
      key="0"
      position={0}
      qaSelector="carusel-item__0"
      src="{{src0}}"
      tags="{{tags0}}"
    />
    <CaruselItem
      key="1"
      position={1}
      qaSelector="carusel-item__1"
      src="{{src1}}"
      tags="{{tags1}}"
    />
    <CaruselItem
      key="2"
      position={0}
      qaSelector="carusel-item__2"
      src="{{src0}}"
      tags="{{tags0}}"
    />
    <CaruselItem
      key="3"
      position={1}
      qaSelector="carusel-item__3"
      src="{{src1}}"
      tags="{{tags1}}"
    />
  </div>
  <div
    className="carusel_controls"
  >
    <div
      className="carusel_control--left"
      data-cy="carusel-controller--left"
      onClick={[Function]}
    >
      Prev
    </div>
    <div
      className="carusel_control--right"
      data-cy="carusel-controller--right"
      onClick={[Function]}
    >
      Next
    </div>
  </div>
</section>
`;
